WIP:  Add marketplace.sql

mvgiambrone opened the pull request:
This is a work in progress, currently trying to duplicate the old marketplace tables in the new database.

Commit:
Add  marketplace.sql

mvgiambrone committed:
Merge branch 'master' into marketplace

mvgiambrone committed:
Added a module for marketplace

mvgiambrone committed:
Fixed user_id column in marketplace_items table

Commit:
Merge branch 'master' into marketplace

Conflicts:
	donut/__init__.py

Commit:
Used base HTML from donut-legacy

tjleing committed:
Merge remote-tracking branch 'origin/master' into marketplace

tjleing committed:
Fixed link from index.html to /marketplace

tjleing committed:
Added routes and some helpers, still not sure how to populate the databases

tjleing committed:
Fixed error

allisonkong commented:
Let me know if you have any questions about marketplace (I'm the original author of the php implementation) or if you want me to review anything :)

tjleing commented:
> Let me know if you have any questions about marketplace (I'm the original author of the php implementation) or if you want me to review anything :)

Awesome; thank you! :)

tjleing committed:
Fixed json parser, meaning that 1/marketplace_items now provides evidence that the code works

tjleing committed:
Created some semblance of the top bar for the marketplace, which has (broken) links to the different categories.

tjleing committed:
Added logic for breaking up the top bar links into rows, and started
moving away from the REST API used for testing the marketplace.

tjleing committed:
Working towards enabling search, also might have to change the url templates a little
Also some UI changes

tjleing committed:
Cleaned up marketplace.html and made it more readable
Comments in both .html templates
Didn't need to pass urls; now just using url_for in the html itself
Added "All Categories" as an option in the dropdown

tjleing committed:
Moved redundant topbar code to its own html file
Started working on page to sell items

tjleing committed:
Beginning of sell.html and helper functions for displaying results correctly

tjleing committed:
Using the helper functions for looking up the ids now

tjleing committed:
Table column titles are working now, as well as showing all from a
single category.  Next up is probably either supporting queries that
are based on what the user enters into the search box, or going back to
the create new listing page.

tjleing committed:
So with NATURAL JOIN the hacky solution I did last commit is now gone
Yay

tjleing committed:
Removed useless prints

tjleing committed:
Making the table for queries look better, with proper headers
and also merged columns for item_title and textbook_title

dqu123 commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,274 @@
+        return edition
+
+
+def get_name_from_user_id(user_id):
```
You can use the `members_full_name` view for this.

tjleing replied:
Fixed.

dqu123 commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,274 @@
+    # want to avoid having exactly 1 category in the remainder row if possible.
+    num_cats = len(categories);
+    num_cols = num_cats
+    if num_cats <= 5:
```
This logic is a bit hard to understand. Is it completely necessary or is there a different way of displaying? I'm also not completely sure what the issue is. Maybe a drawing would help.

tjleing committed:
Beginning of query processing to make search bar work (mostly just ISBN processing)
also process_edition would fail for 11, 12 and 13, but now they should be fine for all numbers

tjleing committed:
Added query tokenizing logic, simplified column logic, used
members_full_name for get_name_from_user_id

tjleing committed:
Added links to elements in the tables generated when searching.  Links
don't do anything yet because I haven't implemented a page to display
the details of an item.

Major refactoring including splitting
helpers.get_marketplace_items_list_data into multiple functions.  The two are together because I needed to refactor in order to implement links.

tjleing committed:
Small refinements to link system before I move on to something else

tjleing committed:
Refactoring helpers.py to centralize queries to the database, and lots
of progress on selling.

tjleing committed:
More progress on selling.  Next up: making the pages actually do things
and converting from session storage to invisible fields (I've rethought
my decision to make them session storage)

tjleing committed:
Added functionality for the rest of sell_10 (textbook select) except for
searching
also general bugfixes

tjleing committed:
Finished sell_2.html, some bugfixes for get_table_list_data, and some
brand new auto_increments in the sql

tjleing committed:
Finalized a preliminary version of creating listings
in this commit: some bugfixes, and actually inserting the listing
into the database
next up: item additional details viewer

tjleing committed:
Implemented item viewing page for more details and also contact info for
the seller.
Next up: getting searching for items actually working

tjleing committed:
Queries work

tjleing committed:
Final touchups before submitting for review

dqu123 commented:
@bluezebragames, is there currently a way to remove items if they've been bought?

tjleing committed:
Fixed crashing typo and added quotes around values in jinja so that
strings with spaces are handled correctly.  Next up: validation.

tjleing commented:
@dqu123 not yet, and moreover, editing is not supported either (although most of the infrastructure for it is up already).  The reasoning was that both of those features are gated behind logging in as either the creator of the item or as admin, and AFAIK privileges aren't working yet.  Also, since I don't have a donut account, testing that would be difficult.  I could make those features accessible to everyone, with the gating being easily addable later; is that preferable?

tjleing committed:
Validation for listings.

tjleing committed:
Took out page numbers that were used mostly for debugging, and added a
section to display how many search results were found.

tjleing commented:
### Summary
- These changes implement the marketplace functionality of donut.
- First, users can look at items and can filter listings by category and/or search queries.
	- The search bar supports filtering by words in the title, textbook author, and ISBN.  Just searching with any of those components will return matches.  Perfect matches are displayed, if any, and if not, imperfect matches are.  Certain words, like 'a,' 'all,' and 'am,' are ignored.
- Also, users can add listings to the database, selecting item condition, price, title, and additional details, and if the item is a textbook, ISBN and author.
- The user can pick the textbook that they are selling from a list, which specifies the title and author automatically, and if the textbook they are selling is not available, they can add it.

### Test Plan
- [ ] Make sure that searching works, by selecting individual categories, and querying by words and ISBNs.
- [x] Make sure that queries that return no results are handled properly (and not just an empty table)
- [x] Make sure that items of category textbook and non-textbook can be added successfully.
- [x] Make sure that new textbooks can be added to the database by using the form.
- [ ] Make sure that just by switching pages (using the next and previous buttons) does not erase any data
	- [ ] even if the data has spaces in it, because that was causing trouble earlier
- [x] Make sure that you cannot continue to the next page of the form if any required data is missing, or if any other data is malformed
- [ ] Make sure that even by going into the HTML and changing the hidden form elements, malformed data cannot be inserted into the database
- [ ] Check for SQL injections by inputting apostrophes into search boxes and form boxes and seeing if the app crashes

dqu123 commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,754 @@
+
+# taken from donut-legacy, which was apparently taken from a CS11
+# C++ assignment by dkong
+skip_words = ["a", "all", "am", "an", "and", "are", "as", "at",
```
Should be 
```python 
SKIP_WORDS = ...
```
see https://www.python.org/dev/peps/pep-0008/#constants

dqu123 commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,754 @@
+    """
+    Turns a string with a query into a list of tokens that represent the query.
+    """
+    global skip_words
```
Is the `global` keyword needed?

dqu123 commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,754 @@
+            # if the tens digit is 1, it's always "th"
+            if (edition / 10) % 10 == 1:
+                return str(edition) + "th"
+            if edition%10 == 1:
```
This is a bit hard to read. Try:
```python
if edition % 10 == 1:
```

dqu123 commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,754 @@
+    # add spaces around method to make the query work
+    method = " " + method + " "
+    # Build the SELECT and FROM clauses
+    s = sqlalchemy.sql.select(fields).select_from(sqlalchemy.text(method.join(tables)))
```
What happens if you run this with something malicious for `method`?

dqu123 commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,754 @@
+        datalist: the original table, but with the two columns merged
+        fields: the column titles similarly merged together into item_title
+    """
+    item_index = -1
```
`None` is more readable. You can check 
```python
if index is None:
```

dqu123 commented on donut/modules/marketplace/routes.py:
```diff
@@ -0,0 +1,342 @@
+        # but if we pass it in, get it
+        page = int(flask.request.form["page"])
+
+    if not page in [1, 10, 2, 3, 4]:
```
Using an enum and adding it to a `constants.py` might be easier.
However, we might want to wait for us to move to Python 3 since that introduces enums natively. 
I'm going to try to get a Python 3 PR merged before this / prio someone get it done
before we merge this PR.

dqu123 commented on donut/modules/marketplace/routes.py:
```diff
@@ -0,0 +1,342 @@
+@blueprint.route('/marketplace/sell', methods=['GET', 'POST'])
+def sell():
+    # if the data or the page in general has errors, we don't let the user continue to the next page
+    hasErrors = False
```
Should be `has_errors`.

dqu123 commented on donut/modules/marketplace/routes.py:
```diff
@@ -0,0 +1,342 @@
+        cat_title=cat_id_map[stored["cat_id"]]
+
+        # full_name and email from user_id
+        data = helpers.get_table_list_data(['members', 'members_full_name'], ['full_name', 'email'], {'user_id': stored['user_id']}, "NATURAL LEFT JOIN")[0]
```
Since you use the `user_id` from stored, which comes from the first result,
this always uses the same `user_id` (not the `user_id` of the user who is logged in).
You should check if there's a helper function to do this in the core module and add it if there's not.

allisonkong commented on donut/modules/marketplace/helpers.py:
```diff
@@ -0,0 +1,754 @@
+import routes
+
+# taken from donut-legacy, which was apparently taken from a CS11
+# C++ assignment by dkong
```
Yeah, I got it from here:
http://users.cms.caltech.edu/~donnie/cs11/cpp/lab7/index.html

It looks like Donnie's completely redone the C++ track since

tjleing committed:
Fixed comments except the one about enums and python3.

tjleing committed:
Fixed one last comment that I forgot about
still not using python3 enums for pages, but will do that after python3
branch is merged

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
2to3 is helpful
